File: data/src/test/java/com/crhonvas/data/mappers/user/UserApiToDomainMapperTest.java
package com.crhonvas.data.mappers.user;

import com.crhonvas.data.remote.model.user.UserApiResponse;
import com.crhonvas.domain.model.user.Address;
import com.crhonvas.domain.model.user.Company;
import com.crhonvas.domain.model.user.User;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Tests for {@link UserApiToDomainMapper}
 */
public class UserApiToDomainMapperTest {

    private static final UserApiResponse userApiResponse = new UserApiResponse.Builder()
            .id(123414)
            .name("name")
            .username("username")
            .email("email")
            .phone("phone")
            .website("website")
            .build();

    @Mock
    ApiAddressToDomainMapper apiAddressToDomainMapper;
    @Mock
    ApiCompanyToDomainMapper apiCompanyToDomainMapper;

    private UserApiToDomainMapper userApiToDomainMapper;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        userApiToDomainMapper = new UserApiToDomainMapper(apiAddressToDomainMapper, apiCompanyToDomainMapper);
    }

    @Test
    public void map() throws Exception {
        when(apiAddressToDomainMapper.map(any())).thenReturn(new Address.Builder().build());
        when(apiCompanyToDomainMapper.map(any())).thenReturn(new Company.Builder().build());
        User actual = userApiToDomainMapper.map(userApiResponse);

        verify(apiAddressToDomainMapper,times(1)).map(any());
        verify(apiCompanyToDomainMapper,times(1)).map(any());
        compare(actual, userApiResponse);
    }

    @Test
    public void map_null_returnEmpty() throws Exception {
        User actual = userApiToDomainMapper.map(null);
        compare(actual, new UserApiResponse.Builder().build());
    }

    private static void compare(User actual, UserApiResponse expected) {
        Assert.assertEquals(expected.getEmail(), actual.getEmail());
        Assert.assertEquals(expected.getName(), actual.getName());
        Assert.assertEquals(expected.getUsername(), actual.getUserName());
        Assert.assertEquals(expected.getId(), actual.getId());
        Assert.assertEquals(expected.getWebsite(), actual.getWebsite());
        Assert.assertEquals(expected.getPhone(), actual.getPhone());
    }
}

File: data/src/test/java/com/crhonvas/data/mappers/user/UserApiToDomainMapperTest.kt
package com.crhonvas.data.mappers.user

import com.crhonvas.data.remote.model.user.UserApiResponse
import com.crhonvas.domain.model.user.Address
import com.crhonvas.domain.model.user.Company
import com.crhonvas.domain.model.user.User
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Matchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations

/**
 * Tests for [UserApiToDomainMapper]
 */
class UserApiToDomainMapperTest {
    @Mock
    var apiAddressToDomainMapper: ApiAddressToDomainMapper? = null

    @Mock
    var apiCompanyToDomainMapper: ApiCompanyToDomainMapper? = null
    private var userApiToDomainMapper: UserApiToDomainMapper? = null
    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        userApiToDomainMapper = UserApiToDomainMapper(apiAddressToDomainMapper!!, apiCompanyToDomainMapper!!)
    }

    @Test
    @Throws(Exception::class)
    fun map() {
        Mockito.`when`(apiAddressToDomainMapper!!.map(Matchers.any())).thenReturn(Address.Builder().build())
        Mockito.`when`(apiCompanyToDomainMapper!!.map(Matchers.any())).thenReturn(Company.Builder().build())
        val actual = userApiToDomainMapper!!.map(userApiResponse)
        Mockito.verify(apiAddressToDomainMapper, Mockito.times(1)).map(Matchers.any())
        Mockito.verify(apiCompanyToDomainMapper, Mockito.times(1)).map(Matchers.any())
        compare(actual, userApiResponse)
    }

    @Test
    @Throws(Exception::class)
    fun map_null_returnEmpty() {
        val actual = userApiToDomainMapper!!.map(null)
        compare(actual, UserApiResponse.Builder().build())
    }

    companion object {
        private val userApiResponse = UserApiResponse.Builder()
            .id(123414)
            .name("name")
            .username("username")
            .email("email")
            .phone("phone")
            .website("website")
            .build()

        private fun compare(actual: User, expected: UserApiResponse) {
            Assert.assertEquals(expected.getEmail(), actual.email)
            Assert.assertEquals(expected.getName(), actual.name)
            Assert.assertEquals(expected.getUsername(), actual.userName)
            Assert.assertEquals(expected.getId(), actual.id)
            Assert.assertEquals(expected.getWebsite(), actual.website)
            Assert.assertEquals(expected.getPhone(), actual.phone)
        }
    }
}